dali/python/nvidia/dali/plugin/pytorch/experimental/proxy/__init__.py

```diff
@@ -0,0 +1,518 @@
+# | main  |   | dali_output_q |   | data_thread | | dali_input_q  |   | worker_0  | | worker_1  |
+# +-------+   +---------------+   +-------------+ +---------------+   +-----------+ +-----------+
+
+
+def _external_source_node_names(pipeline):
+    """
+    extract the names of all the ExternalSource nodes in the pipeline
+    """
+    # TODO(janton): Add a native function to query those names, so that we can do it
+    # also on deserialized pipelines
+    if pipeline._deserialized:
+        raise RuntimeError(
+            "Not able to find the external source "
+            "operator names, since the pipeline was deserialized"
+        )
+    if not pipeline._py_graph_built:
+        pipeline._build_graph()
+    input_node_names = []
+    for op in pipeline._ops:
+        if isinstance(op._op, ExternalSource):
+            input_node_names.append(op.name)
+    return input_node_names
+
+
+class DALIPipelineRunRef:
+    """
+    Reference for a DALI pipeline run iteration.
+    """
+
+    def __init__(self, batch_id, inputs):
+        """
+        batch_id: Identifier of the batch
+        is_scheduled: Whether the iteration has been scheduled for execution already
+        inputs: Inputs to be used when scheduling the iteration (makes sense only if
+        is_scheduled is False)
+        """
+        self.batch_id = batch_id
+        self.inputs = inputs
+        assert self.inputs is not None
+        self.is_scheduled = False
+
+
+class DALIProcessedSampleRef:
+    """
+    Placeholder for a pipeline run reference, which is returned by the data worker instead of
+    the actual data. The PyTorch worker returns this trivial object, only containing information
+    about this proxy instance and the input data to the pipeline. Later in the collate function,
+    we send the data for execution to DALI.
+    """
+
+    def __init__(self, proxy, inputs):
+        self.proxy = proxy
+        self.inputs = inputs
+        if len(inputs) != len(self.proxy._dali_input_names):
+            raise RuntimeError(
+                f"Unexpected number of inputs. Expected: {self._dali_input_names}, got: {inputs}"
+            )
+
+
+def _collate_dali_processed_sample_ref_fn(samples, *, collate_fn_map=None):
+    """
+    Special collate function that schedules a DALI iteration for execution
+    """
+    assert len(samples) > 0
+    sample = samples[0]
+    inputs = [[] for _ in range(len(sample.inputs))]
+    proxy = sample.proxy
+    for sample in samples:
+        assert proxy == sample.proxy
+        for idx, input_ref in enumerate(sample.inputs):
+            inputs[idx].append(input_ref)
+    pipe_run_ref = proxy._create_pipe_run_ref(inputs)
+    if not proxy._deterministic:
+        proxy._schedule_batch(pipe_run_ref)
+        # No need for the inputs now
+        pipe_run_ref.inputs = None
+        # Mark as already scheduled
+        pipe_run_ref.is_scheduled = True
```

Comment:
:+1: for NOT relying on a secret proxy value (like `_deterministic`) at the receiving end, but rather having this explicit.